src/gui/qgsfieldmappingmodel.h

```diff
@@ -0,0 +1,165 @@
+/***************************************************************************
+  qgsfieldmappingmodel.h - QgsFieldMappingModel
+
+ ---------------------
+ begin                : 17.3.2020
+ copyright            : (C) 2020 by Alessandro Pasotti
+ email                : elpaso at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSFIELDMAPPINGMODEL_H
+#define QGSFIELDMAPPINGMODEL_H
+
+#include <QAbstractTableModel>
+#include <QStyledItemDelegate>
+
+#include "qgsfields.h"
+#include "qgsexpression.h"
+#include "qgsexpressioncontextgenerator.h"
+#include "qgsfieldconstraints.h"
+#include "qgis_gui.h"
+
+
+/**
+ * \ingroup gui
+ * The QgsFieldMappingModel holds mapping information for mapping from one set of QgsFields to another,
+ * for each set of "destination" fields an expression defines how to obtain the values of the
+ * "destination" fields.
+ * The model can be optionally set "editable" allowing to modify all the fields, by default only
+ * the mapping expression is editable.
+ *
+ * \since QGIS 3.14
+ */
+class GUI_EXPORT QgsFieldMappingModel: public QAbstractTableModel
+{
+
+    Q_OBJECT
+
+  public:
+
+    /**
+     * The ColumnDataIndex enum represents the column index for the view
+     */
+    enum class ColumnDataIndex : int
+    {
+      SourceExpression,       //!< Expression
+      DestinationName,        //!< Destination field name
+      DestinationType,        //!< Destination field QVariant::Type casted to (int)
+      DestinationLength,      //!< Destination field length
+      DestinationPrecision,   //!< Destination field precision
+      DestinationConstraints, //!< Destination field constraints
+    };
+
+    Q_ENUM( ColumnDataIndex );
+
+    /**
+     * The Field struct holds information about a mapped field
+     */
+    struct Field
+    {
+      //! The original name of the field
+      QString originalName;
+      //! The field in its current status (it might have been renamed)
+      QgsField field;
+      //! The expression for the mapped field from the source fields
+      QgsExpression expression;
```

Comment:
Does this need to be a QgsExpression? As far as I can see we could get away with a QString alone, and avoid the unnecessary expression parsing.

Reply:
Yes, I though about it and my rationale was the perhaps having a QgsExpression would have spared additional conversions when checking for validity. But I'm ok with using a string here.